### PR TITLE
Merge Labled dataset

### PR DESCRIPTION
**Pull request by jannisborn**
added abstract class for labeled dataset, called `AnnotatedDataset`, found under `pytoda.datasets.annotated_dataset`:

- receives  a `torch.utils.data.dataset` object, e.g. a previously instantiated child object of type `SMILESDataset`
- receives a path to a `csv` file with labels. The labels can be multi task but since the amount of tasks is dynamically inferred, the labels in a multi class (but single task) case (like MNIST) *should not be defined as one-hot vectors*.
- The last column of the label-csv file needs to contain IDs that are also contained in the associated `dataset` object (similar to `drug_sensitivity_dataset`)

Yapfed some other files on the fly.

@drugilsberg, not sure this can handle SMILES augmentation on the fly. If so, it should be done by `sample_to_index_mapping`. Any idea on this?

### CODE REVIEW

**Comment by drugilsberg** [comment]
@jannisborn the SMILESDataset class should handle it inherently.

**Comment by jannisborn** [comment]
great, I just double checked, you're right @drugilsberg! Think we can merge then

File: pytoda/datasets/annotated_dataset.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        Dataset.__init__(self)
+        self.annotations_filepath = annotations_filepath
+        # device
+        self.device = device
+        # backend
+        self.backend = backend
+        # e.g. SMILES
+        self.input_data = input_data
+
+        self.annotated_data_df = pd.read_csv(
```
Comment (drugilsberg) [inline pytoda/datasets/annotated_dataset.py:45]: Here I would either include `index_col` in `kwargs` or make sure to add `_ = kwargs.pop('index_col', None)` before `pd.read_csv` to make sure we don't raise.

File: pytoda/datasets/annotated_dataset.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        device('cuda' if torch.cuda.is_available() else 'cpu'),
+        backend: str = 'eager',
+        **kwargs
+    ) -> None:
+        """
+        Initialize an annotated dataset.
+        E.g. the  input_data could be SMILES and the annotations could be
+        single or multi task labels.
+
+        Args:
```
Comment (drugilsberg) [inline pytoda/datasets/annotated_dataset.py:26]: Updated the docstring according to the arguments. In case we "pop" index_col from kwargs add the fact that this argument should not be provided in the docstring.

File: pytoda/datasets/annotated_dataset.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+class AnnotatedDataset(Dataset):
+    """
+    Annotated dataset implementation.
+    """
+
+    def __init__(
+        self,
+        annotations_filepath: str,
+        input_data: Dataset,
```
Comment (drugilsberg) [inline pytoda/datasets/annotated_dataset.py]: Remove `backend` since it's unsed. I would rename `input_data` to `dataset` for consistency.

**Review by drugilsberg — CHANGES_REQUESTED** [review]

File: pytoda/datasets/annotated_dataset.py (hunk 3 above)
Comment (jannisborn) [reply pytoda/datasets/annotated_dataset.py]: both done

File: pytoda/datasets/annotated_dataset.py (hunk 1 above)
Comment (jannisborn) [reply pytoda/datasets/annotated_dataset.py:45]: I updated the default of `index_col` (which is `None`) to 0 (set in constructor)

File: pytoda/datasets/annotated_dataset.py (hunk 2 above)
Comment (jannisborn) [reply pytoda/datasets/annotated_dataset.py:26]: should also be fixed with latest commit

**Review by jannisborn — COMMENTED** [review]
should all be fixed now

**Review by drugilsberg — APPROVED** [review]
